android/engine/src/main/java/org/smartregister/fhircore/engine/rulesengine/RulesExecutor.kt

```diff
@@ -0,0 +1,208 @@
+import org.smartregister.fhircore.engine.domain.model.RuleConfig
+import org.smartregister.fhircore.engine.domain.model.ViewType
+import org.smartregister.fhircore.engine.util.extension.extractLogicalIdUuid
+
+class RulesExecutor @Inject constructor(val rulesFactory: RulesFactory) {
+
+  suspend fun processResourceData(
+    baseResource: Resource,
+    relatedRepositoryResourceData: LinkedList<RepositoryResourceData>,
+    ruleConfigs: List<RuleConfig>
+  ): ResourceData {
+    val relatedResourcesMap = relatedRepositoryResourceData.createRelatedResourcesMap()
+    val computedValuesMap =
+      computeRules(
+        ruleConfigs = ruleConfigs,
+        baseResource = baseResource,
+        relatedResourcesMap = relatedResourcesMap
+      )
+    return ResourceData(
+      baseResourceId = baseResource.logicalId.extractLogicalIdUuid(),
+      baseResourceType = baseResource.resourceType,
+      computedValuesMap = computedValuesMap
+    )
+  }
+
+  /**
+   * This function pre-computes all the Rules for [ViewType]'s of List including list nested in the
+   * views. The LIST view computed values includes the parent's.
+   */
+  suspend fun processListResourceData(
+    listProperties: ListProperties,
+    relatedRepositoryResourceData: LinkedList<RepositoryResourceData>,
+    computedValuesMap: Map<String, Any>
+  ): List<ResourceData> {
+    val relatedResourcesMap = relatedRepositoryResourceData.createRelatedResourcesMap()
+    return listProperties.resources.flatMap { listResource ->
+      filteredListResources(relatedResourcesMap, listResource)
+        .mapToResourceData(
+          relatedResourcesMap = relatedResourcesMap,
+          ruleConfigs = listProperties.registerCard.rules,
+          listRelatedResources = listResource.relatedResources,
+          computedValuesMap = computedValuesMap
+        )
+    }
+  }
+
+  private suspend fun computeRules(
+    ruleConfigs: List<RuleConfig>,
+    baseResource: Resource,
+    relatedResourcesMap: Map<String, List<Resource>>
+  ): Map<String, Any> =
+    // Compute values via rules engine and return a map. Rule names MUST be unique
+    rulesFactory.fireRules(
+      rules = rulesFactory.generateRules(ruleConfigs),
+      baseResource = baseResource,
+      relatedResourcesMap = relatedResourcesMap
+    )
+
+  private suspend fun List<Resource>.mapToResourceData(
+    relatedResourcesMap: MutableMap<String, MutableList<Resource>>,
+    ruleConfigs: List<RuleConfig>,
+    listRelatedResources: List<ExtractedResource>,
+    computedValuesMap: Map<String, Any>
+  ) =
+    this.map { resource ->
+      val listItemRelatedResources: Map<String, List<Resource>> =
+        listRelatedResources.associate { (id, resourceType, fhirPathExpression) ->
+          (id
+            ?: resourceType.name) to
+            rulesFactory.rulesEngineService.retrieveRelatedResources(
+              resource = resource,
+              relatedResourceKey = id ?: resourceType.name,
```

Comment:
so the constraint here is that you have to pass an ID or you can't have multiple listRelatedResources with the same resourceType. I think it would be nice to throw an error here if we see a id or resourceType.name that is duplicated

Reply:
I'm guessing this is so that we warn the user that they misconfigured with at duplicate?? If so I think an exception would be most useful to the writer of the configs, otherwise by the time the exception shows on the app it is probably too late. 

Once we implement a configs validator such a case should not happen and the issue would have been caught before hand

If we still need to throw/log such an exception let me know and we can refactor to accommodate this

Reply:
yea exactly, I think it's better to throw here and catch with a message somewhere higher up the call stack, even if we validate configs we're going to run this against arbitrary FHIR APIs (eventually) so we can't assume the config is valid

Reply:
resolving since this will get thrown by retrieveRelatedResources